golang/dep
# make importers error tolerant

### What does this do / why do we need it?

This makes the import process much more error tolerant. In case of a catastrophic failure, a warning is logged and the root analyzer gives up on importing the external config but _only for that directory_. On the other hand, for less serious errors such as failure to load `glide.lock` etc., no error is returned and only a warning is logged.

### What should your reviewer look out for in this PR?

Is the approach in the correct direction? There's still work left to be done but I would like to get some early feedback.

### Do you need help or clarification on anything?

- While this behavior of bad config tolerance is a good default, I think there should be a way to turn it off if required. Do you agree with this and if so, should it be a part of this PR or a separate one? It might make sense to club this with the possibility of having a limit on the number of recoverable errors before erroring out.
- Any tips for writing the tests? There are a _lot_ of scenarios from what I can tell.

### Which issue(s) does this PR fix?

fixes #1315

<details>
<summary>Code changelog</summary>

- root_analyzer.go: Log a warning on encountering an unrecoverable error during
  import of external config and proceed with the import for other packages.
  Do not return error from private func `importManifestAndLock`.

`internal/importers`:
- base/importer.go:
  - `loadPackages`: Do not return error. If `SourceManager.DeduceProjectRoot`
    fails to determine the project root, log a warning and continue with the
    rest of the imported packages.
  - `ImportPackages`: Do not return error. When constraint resolution from the
    lock hint fails, only log a warning.
- Importer implementations:
  - Return an error from `Import` only for catastrophic failures(ex: yaml
    parsing failed).
  - `load`: Make it more error tolerant. Log warnings only for any of the
    following scenarios:
    - When and if a lock file, separate from the dependency file is present,
      like, in the case of glide, and parsing fails. Continue with the import
      as if the lock file was not present.
    - If import packages are parsed line by line like in the case of glock,
      and one of the line could not be parsed.
  - `convert`: Do not return an error. Log warnings only for any of the
    following scenarios:
    - Expected field, such as `package` for an entry in `glide.yaml>imports`
      is not present.
    - Package was specified but the contraint could not be found.

`tests`: 
- Updated existing importer tests.
- Added tests for importer failure scenarios:
  - `internal/importers/base/importer_test.go`:
    - Check for warning when an invalid project is present whose project root
      cannot be parsed.
    - Check for warning when lock hint cannot be resolved correctly and the 
      constraint cannot be applied.
  - Integration test for malformed external config(glide.yaml)

</details>
<br>

**TODO:**
- [X] Add and update tests
